exonum-light-client/src/main/java/com/exonum/client/ExonumHttpClient.java

```diff
@@ -141,19 +149,76 @@ public BlockResponse getBlockByHeight(long height) {
   }
 
   @Override
-  public BlocksResponse getBlocks(int count, BlockFilteringOption blockFilter, long heightMax,
+  public BlocksRange getBlocks(long fromHeight, long toHeight, BlockFilteringOption blockFilter,
       BlockTimeOption timeOption) {
-    checkArgument(0 < count,
-        "Requested number of blocks should be positive number but was %s", count);
-    return doGetBlocks(count, blockFilter, heightMax, timeOption);
+    checkArgument(0 <= fromHeight, "First block height (%s) must be non-negative", fromHeight);
+    checkArgument(fromHeight <= toHeight,
+        "First block height (%s) should be less than or equal to the last block height (%s)",
+        fromHeight, toHeight);
+
+    // 'maximum' as when skipping empty the actual might be way smaller
+    int maxSize = Math.toIntExact(toHeight - fromHeight + 1);
+    List<Block> blocks = new ArrayList<>(maxSize);
+    for (long rangeLast = toHeight; rangeLast >= fromHeight; ) {
+      int remainingBlocks = Math.toIntExact(rangeLast - fromHeight + 1);
+      int numBlocks = min(remainingBlocks, MAX_BLOCKS_PER_REQUEST);
+      BlocksResponse blocksResponse = doGetBlocks(numBlocks, blockFilter, rangeLast, timeOption);
+
+      blocks.addAll(blocksResponse.getBlocks());
+
+      rangeLast = blocksResponse.getBlocksRangeStart() - 1;
+    }
+
+    return postProcessResponseBlocks(fromHeight, toHeight, blocks);
   }
 
   @Override
-  public BlocksResponse getLastBlocks(int count, BlockFilteringOption blockFilter,
+  public BlocksRange getLastBlocks(int count, BlockFilteringOption blockFilter,
       BlockTimeOption timeOption) {
     checkArgument(0 < count,
         "Requested number of blocks should be positive number but was %s", count);
-    return doGetBlocks(count, blockFilter, null, timeOption);
+
+    List<Block> blocks = new ArrayList<>(count);
+    // The first request does not specify the maximum height to get the top blocks
+    long blockchainHeight = Long.MIN_VALUE;
+    Long nextHeight = null;
+    int remainingBlocks = count;
+    while (remainingBlocks > 0
+        && (nextHeight == null || nextHeight >= GENESIS_BLOCK_HEIGHT)) {
+      int numBlocks = min(remainingBlocks, MAX_BLOCKS_PER_REQUEST);
+      BlocksResponse blocksResponse = doGetBlocks(numBlocks, blockFilter, nextHeight, timeOption);
+
+      blocks.addAll(blocksResponse.getBlocks());
+
+      nextHeight = blocksResponse.getBlocksRangeStart() - 1;
+      blockchainHeight = max(blockchainHeight, blocksResponse.getBlocksRangeEnd() - 1);
+      remainingBlocks = Math.toIntExact(count - (blockchainHeight - nextHeight));
+    }
+
+    long fromHeight = max(blockchainHeight - count + 1, GENESIS_BLOCK_HEIGHT);
```

Comment:
looks like it's the wrong statement. For example, we need 10 non-empty blocks in the range 50..100 height, but blocks 91..95 are empty.
`fromHeight` in this case will be 100-10 + 1 = 91.
But the correct one is 86

Reply:
It is correct indeed — please see the updated spec of this method. It is specified to return _up to_ min(count, blockchainHeight + 1) blocks. Filtering no longer affects the range. E.g., if you request top 10 excluding empty, you will get top 10 excluding empty.

Reply:
> The loop is finished at this point, but blocks.size() is 1000, not 1100.

That is correct by the spec — because it says it will return the correctly sized range of blocks, it will not extend it arbitrarily.

There are two kind of things one might expect from this method: return top N blocks _filtering_ empty; or return top N _non-empty_ blocks. These are different requests. This method allows to request a sized _range_ of blocks, not a certain number of (non-empty) blocks.

The fact that we are having this confusion probably means this API is a monstrosity and we will have to discuss it in a more detail in jira.

Reply:
@bullet-tooth @MakarovS @DENMROOT , please see https://jira.bf.local/browse/ECR-3255?focusedCommentId=48525&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-48525